crypto/crypto.h

```diff
@@ -0,0 +1,121 @@
+#ifndef SECVARCTL_CRYPTO_H
+#define SECVARCTL_CRYPTO_H
+
+#ifdef OPENSSL
+
+#include <openssl/obj_mac.h>
+#define CRYPTO_MD_SHA1 NID_sha1
+#define CRYPTO_MD_SHA224 NID_sha224
+#define CRYPTO_MD_SHA256 NID_sha256 
+#define CRYPTO_MD_SHA384 NID_sha384
+#define CRYPTO_MD_SHA512 NID_sha512
+
+#else
+
+#include <mbedtls/md.h>
+#define CRYPTO_MD_SHA1 MBEDTLS_MD_SHA1
+#define CRYPTO_MD_SHA224 MBEDTLS_MD_SHA224
+#define CRYPTO_MD_SHA256 MBEDTLS_MD_SHA256 
+#define CRYPTO_MD_SHA384 MBEDTLS_MD_SHA384
+#define CRYPTO_MD_SHA512  MBEDTLS_MD_SHA512
+
+#endif
+/**====================PKCS7 Functions ====================**/
+
+/* 
+ *checks the pkcs7 struct for using SHA256 as the message digest 
+ *@param pkcs7 , a pointer to either an openssl or mbedtls pkcs7 struct
+ *@return SUCCESS if message digest is SHA256 else return errno
+ */
+int crypto_pkcs7_is_md_sha256(void *pkcs7);
+
+/*
+ *free's the memory allocated for a pkcs7 structure
+ *@param pkcs7 , a pointer to either an openssl or mbedtls pkcs7 struct
+ */
+void crypto_pkcs7_free(void *pkcs7);
+
+/*
+ *parses a buffer into a pointer to a pkcs7 struct. struct allocation is done internally to this func, but not dealloc
+ *@param buf, buffer of data containg pkcs7 data
+ *@param buflen, length of buf
+ *@return if successful, a void pointer to either an openssl or mbedtls pkcs7 struct. else returns NULL
+ *NOTE: if successful (returns not NULL), remember to call crypto_free_pkcs7 to unalloc. 
+ */
+void * crypto_pkcs7_parse_der(const unsigned char *buf, const int buflen);
+
+/*
+ *returns one signing ceritficate from the PKKCS7 signing certificate chain
+ *@param pkcs7 ,  a pointer to either an openssl or mbedtls pkcs7 struct
+ *@param cert_num , the index (starts at 0) of the signing certificate to retrieve
+ *@return a pointer to either an openssl or mbedtls X509 Struct
+ */
+void* crypto_get_signing_cert(void *pkcs7, int cert_num);
+
+
+/**====================X509 Functions ====================**/
+int crypto_get_x509_der_len(void *x509);
+int crypto_get_tbs_x509_der_len(void *x509);
+int crypto_get_x509_version(void *x509);
+int crypto_get_x509_sig_len(void *x509);
+int crypto_x509_md_is_sha256(void *x509);
+int crypto_x509_oid_is_pkcs1_sha256(void *x509);
+int crypto_x509_get_pk_bit_len(void *x509);
```

Comment:
Are these `void *x509` pointers usually to structs that contain their sizes, or is there a possibility these might point to some kind of arbitrary allocated memory? If mbedtls/openssl know how big their data is supposed to be, then I'm fine with this.

I think its just the `void *` that is spooking me a little bit. Could be worth defining a type for these that can be handled in the `#ifdef`s, but that might just be overkill.

Reply:
Up to you guys. I am not sure what you mean by "structs that contain their sizes", aren't all structs some fixed size? 
While I don't think that typedef-ing adds any change in the compiled code, I think it adds clarity so I will add that now. Let me know if I should do anything else with regards to this

Reply:
Leave it for now, we can maybe change it in the future if it's a problem.